src/commonMain/kotlin/fr/acinq/lightning/payment/PaymentRequest.kt

```diff
@@ -1,14 +1,28 @@
 package fr.acinq.lightning.payment
 
 import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.bitcoin.utils.Try
+import fr.acinq.bitcoin.utils.runTrying
 import fr.acinq.lightning.Features
 import fr.acinq.lightning.MilliSatoshi
 import fr.acinq.lightning.utils.currentTimestampSeconds
 
-sealed interface PaymentRequest {
-    val amount: MilliSatoshi?
-    val paymentHash: ByteVector32
-    val features: Features
+sealed class PaymentRequest {
+    abstract val amount: MilliSatoshi?
+    abstract val paymentHash: ByteVector32
+    abstract val features: Features
 
-    fun isExpired(currentTimestampSeconds: Long = currentTimestampSeconds()): Boolean
+    abstract fun isExpired(currentTimestampSeconds: Long = currentTimestampSeconds()): Boolean
+
+    abstract fun write(): String
+
+    companion object {
+        fun read(input: String): Try<PaymentRequest> = runTrying {
```

Comment:
Why the `runTrying` since `Bolt11Invoice.read` and `Bolt12Invoice.fromString` already return a `Try`? This is unnecessary, in the exception case you are now explicitly throwing then re-catching the exception, which is extremely inefficient?

Reply:
Actually, I've found a way to make it both readable and efficient with `.map { it }`.